app/controllers/CyaMaintainProfileController.scala

```diff
@@ -191,26 +202,72 @@ class CyaMaintainProfileController @Inject() (
-            val updatedProfile = traderProfile.copy(niphlNumber = niphlNumber)
-            auditService.auditMaintainProfile(traderProfile, updatedProfile, request.affinityGroup)
-            for {
-              _ <- traderProfileConnector.submitTraderProfile(updatedProfile, request.eori)
-            } yield Redirect(navigator.nextPage(CyaMaintainProfilePage, NormalMode, request.userAnswers))
-          }
-        case Left(errors)       =>
-          Future.successful(logErrorsAndContinue(errorMessage, routes.ProfileController.onPageLoad(), errors))
+      (for {
+        niphlNumber      <- handleValidateError(TraderProfile.validateNiphlNumber(request.userAnswers))
+        oldTraderProfile <- traderProfileConnector.getTraderProfile(request.eori)
+        newTraderProfile <- Future.successful(oldTraderProfile.copy(niphlNumber = niphlNumber))
+        _                 = auditService.auditMaintainProfile(oldTraderProfile, newTraderProfile, request.affinityGroup)
+        _                <- submitTraderProfileIfValueChanged(newTraderProfile, oldTraderProfile, NiphlNumberUpdatePage, request.eori)
+      } yield Redirect(navigator.nextPage(CyaMaintainProfilePage, NormalMode, request.userAnswers))
+        .addingToSession(
+          dataUpdated -> isValueChanged(newTraderProfile, oldTraderProfile, NiphlNumberUpdatePage).toString
+        )
+        .addingToSession(dataAdded -> isValueAdded(newTraderProfile, oldTraderProfile, NiphlNumberUpdatePage).toString)
+        .addingToSession(pageUpdated -> niphlNumberUpdatePage)).recover { case e: TraderProfileBuildFailure =>
+        logErrorsAndContinue(e.getMessage, routes.ProfileController.onPageLoad())
       }
   }
 
+  private def isValueChanged(newTraderProfile: TraderProfile, oldTraderProfile: TraderProfile, page: Page): Boolean =
+    page match {
+      case UkimsNumberUpdatePage                      => oldTraderProfile.ukimsNumber != newTraderProfile.ukimsNumber
+      case NirmsNumberUpdatePage | HasNirmsUpdatePage => oldTraderProfile.nirmsNumber != newTraderProfile.nirmsNumber
+      case NiphlNumberUpdatePage | HasNiphlUpdatePage => oldTraderProfile.niphlNumber != newTraderProfile.niphlNumber
+    }
+
+  private def isValueAdded(newTraderProfile: TraderProfile, oldTraderProfile: TraderProfile, page: Page): Boolean =
+    page match {
+      case NirmsNumberUpdatePage => oldTraderProfile.nirmsNumber.isEmpty && newTraderProfile.nirmsNumber.isDefined
+      case NiphlNumberUpdatePage => oldTraderProfile.niphlNumber.isEmpty && newTraderProfile.niphlNumber.isDefined
+    }
+
+  private def isValueRemoved(newTraderProfile: TraderProfile, oldTraderProfile: TraderProfile, page: Page): Boolean =
+    page match {
+      case HasNirmsUpdatePage => oldTraderProfile.nirmsNumber.isDefined && newTraderProfile.nirmsNumber.isEmpty
+      case HasNiphlUpdatePage => oldTraderProfile.niphlNumber.isDefined && newTraderProfile.niphlNumber.isEmpty
+    }
+
+  private def submitTraderProfileIfValueChanged(
```

Comment:
we now only submit if something has changed to cut down on api calls, but we audit regardless